File: ProjectTrackervb/ProjectTracker.Library/ResourceAssignment.vb
<Serializable()> _
Public Class ResourceAssignment
  Inherits BusinessBase(Of ResourceAssignment)

  Implements IHoldRoles

#Region " Business Methods "

  Private _timestamp(7) As Byte

  Private Shared ProjectIdProperty As PropertyInfo(Of Guid) = RegisterProperty(New PropertyInfo(Of Guid)("ProjectId", "Project id", Guid.Empty))
  Private _projectId As Guid = ProjectIdProperty.DefaultValue
  Public ReadOnly Property ProjectId() As Guid
    Get
      Return GetProperty(Of Guid)(ProjectIdProperty, _projectId)
    End Get
  End Property

  Private Shared ProjectNameProperty As PropertyInfo(Of String) = RegisterProperty(New PropertyInfo(Of String)("ProjectName"))
  Private _projectName As String = ProjectNameProperty.DefaultValue
  Public ReadOnly Property ProjectName() As String
    Get
      Return GetProperty(Of String)(ProjectNameProperty, _projectName)
    End Get
  End Property

  Private Shared AssignedProperty As PropertyInfo(Of SmartDate) = RegisterProperty(New PropertyInfo(Of SmartDate)("Assigned"))
  Private _assigned As New SmartDate(Today)
  Public ReadOnly Property Assigned() As String
    Get
      Return GetProperty(Of SmartDate, String)(AssignedProperty, _assigned)
    End Get
  End Property

  Private Shared RoleProperty As PropertyInfo(Of Integer) = RegisterProperty(New PropertyInfo(Of Integer)("Role"))
  Private _role As Integer = RoleProperty.DefaultValue
  Public Property Role() As Integer Implements IHoldRoles.Role
    Get
      Return GetProperty(Of Integer)(RoleProperty, _role)
    End Get
    Set(ByVal value As Integer)
      SetProperty(Of Integer)(RoleProperty, _role, value)
    End Set
  End Property

  Public Function GetProject() As Project

    CanExecuteMethod("GetProject", True)
    Return Project.GetProject(_projectId)

  End Function

  Public Overrides Function ToString() As String

    Return _projectId.ToString

  End Function

#End Region

#Region " Validation Rules "

  Protected Overrides Sub AddBusinessRules()

    ValidationRules.AddRule(AddressOf Assignment.ValidRole, RoleProperty)

  End Sub

#End Region

#Region " Authorization Rules "

  Protected Overrides Sub AddAuthorizationRules()

    AuthorizationRules.AllowWrite(RoleProperty, "ProjectManager")

  End Sub

#End Region

#Region " Factory Methods "

  Friend Shared Function NewResourceAssignment( _
    ByVal projectId As Guid) As ResourceAssignment

    Return DataPortal.CreateChild(Of ResourceAssignment)(projectId, RoleList.DefaultRole)

  End Function

  Friend Shared Function GetResourceAssignment( _
    ByVal data As ProjectTracker.DalLinq.Assignment) As ResourceAssignment

    Return DataPortal.FetchChild(Of ResourceAssignment)(data)

  End Function

  Private Sub New()
    ' require use of factory methods
  End Sub

  Private Sub New(ByVal project As Project, ByVal role As Integer)

    MarkAsChild()
    _projectId = project.Id
    _projectName = project.Name
    _assigned.Date = Assignment.GetDefaultAssignedDate
    _role = role

  End Sub

#End Region

#Region " Data Access "

  Private Overloads Sub Child_Create(ByVal projectId As Guid, ByVal role As Integer)

    Dim proj = Project.GetProject(projectId)
    _projectId = proj.Id
    _projectName = proj.Name
    _assigned.Date = Assignment.GetDefaultAssignedDate
    _role = role

  End Sub

  Private Sub Child_Fetch(ByVal data As ProjectTracker.DalLinq.Assignment)

    _projectId = data.ProjectId
    _projectName = data.Project.Name
    _assigned = data.Assigned
    _role = data.Role
    _timestamp = data.LastChanged.ToArray

  End Sub

  Private Sub Child_Insert(ByVal resource As Resource)

    _timestamp = Assignment.AddAssignment( _
      _projectId, resource.Id, _assigned, _role)

  End Sub

  Private Sub Child_Update(ByVal resource As Resource)

    _timestamp = Assignment.UpdateAssignment( _
      _projectId, resource.Id, _assigned, _role, _timestamp)

  End Sub

  Private Sub Child_DeleteSelf(ByVal resource As Resource)

    Assignment.RemoveAssignment(_projectId, resource.Id)

  End Sub

#End Region

End Class

File: ProjectTrackercs/ProjectTracker.Library/ResourceAssignment.cs
using Csla;
using System;

namespace ProjectTracker.Library
{
  [Serializable()]
  public class ResourceAssignment : BusinessBase<ResourceAssignment>, IHoldRoles
  {
    #region  Business Methods

    private byte[] _timestamp = new byte[8];

    private static PropertyInfo<Guid> ProjectIdProperty = RegisterProperty(new PropertyInfo<Guid>("ProjectId", "Project id", Guid.Empty));
    private Guid _projectId = ProjectIdProperty.DefaultValue;
    public Guid ProjectId
    {
      get
      {
        return GetProperty<Guid>(ProjectIdProperty, _projectId);
      }
    }

    private static PropertyInfo<string> ProjectNameProperty = RegisterProperty(new PropertyInfo<string>("ProjectName"));
    private string _projectName = ProjectNameProperty.DefaultValue;
    public string ProjectName
    {
      get
      {
        return GetProperty<string>(ProjectNameProperty, _projectName);
      }
    }

    private static PropertyInfo<SmartDate> AssignedProperty = RegisterProperty(new PropertyInfo<SmartDate>("Assigned"));
    private SmartDate _assigned = new SmartDate(System.DateTime.Today);
    public string Assigned
    {
      get
      {
        return GetProperty<SmartDate, string>(AssignedProperty, _assigned);
      }
    }

    private static PropertyInfo<int> RoleProperty = RegisterProperty(new PropertyInfo<int>("Role"));
    private int _role = RoleProperty.DefaultValue;
    public int Role
    {
      get
      {
        return GetProperty<int>(RoleProperty, _role);
      }
      set
      {
        SetProperty<int>(RoleProperty, ref _role, value);
      }
    }

    public Project GetProject()
    {
      CanExecuteMethod("GetProject", true);
      return Project.GetProject(_projectId);
    }

    public override string ToString()
    {
      return _projectId.ToString();
    }

    #endregion

    #region  Validation Rules

    protected override void AddBusinessRules()
    {
      ValidationRules.AddRule(Assignment.ValidRole, RoleProperty);
    }

    #endregion

    #region  Authorization Rules

    protected override void AddAuthorizationRules()
    {
      AuthorizationRules.AllowWrite(RoleProperty, "ProjectManager");
    }

    #endregion

    #region  Factory Methods

    internal static ResourceAssignment NewResourceAssignment(Guid projectId)
    {
      return DataPortal.CreateChild<ResourceAssignment>(projectId, RoleList.DefaultRole());
    }

    internal static ResourceAssignment GetResourceAssignment(ProjectTracker.DalLinq.Assignment data)
    {
      return DataPortal.FetchChild<ResourceAssignment>(data);
    }

    private ResourceAssignment()
    { /* require use of factory methods */ }

    #endregion

    #region  Data Access

    private void Child_Create(Guid projectId, int role)
    {
      var proj = Project.GetProject(projectId);
      _projectId = proj.Id;
      _projectName = proj.Name;
      _assigned.Date = Assignment.GetDefaultAssignedDate();
      _role = role;
    }

    private void Child_Fetch(ProjectTracker.DalLinq.Assignment data)
    {
      _projectId = data.ProjectId;
      _projectName = data.Project.Name;
      _assigned = data.Assigned;
      _role = data.Role;
      _timestamp = data.LastChanged.ToArray();
    }

    private void Child_Insert(Resource resource)
    {
      _timestamp = Assignment.AddAssignment(_projectId, resource.Id, _assigned, _role);
    }

    private void Child_Update(Resource resource)
    {
      _timestamp = Assignment.UpdateAssignment(_projectId, resource.Id, _assigned, _role, _timestamp);
    }

    private void Child_DeleteSelf(Resource resource)
    {
      Assignment.RemoveAssignment(_projectId, resource.Id);
    }

    #endregion

  }
}